drivers/serial/eusart_gecko.c

```diff
@@ -0,0 +1,532 @@
+
+static inline enum uart_config_data_bits eusart_gecko_ll2cfg_databits(EUSART_Databits_TypeDef db,
+								      EUSART_Parity_TypeDef p)
+{
+	switch (db) {
+	case eusartDataBits7:
+		if (p == eusartNoParity) {
+			return UART_CFG_DATA_BITS_7;
+		} else {
+			return UART_CFG_DATA_BITS_6;
+		}
+	case eusartDataBits9:
+		if (p == eusartNoParity) {
+			return UART_CFG_DATA_BITS_9;
+		} else {
+			return UART_CFG_DATA_BITS_8;
+		}
+	case eusartDataBits8:
+	default:
+		if (p == eusartNoParity) {
+			return UART_CFG_DATA_BITS_8;
+		} else {
+			return UART_CFG_DATA_BITS_7;
+		}
+	}
+}
+
+/**
+ * @brief  Get LL hardware flow control define from
+ *         Zephyr hardware flow control option.
+ * @note   Supports only UART_CFG_FLOW_CTRL_RTS_CTS and UART_CFG_FLOW_CTRL_RS485.
+ * @param  fc: Zephyr hardware flow control option.
+ * @retval eusartHwFlowControlCtsAndRts, or eusartHwFlowControlNone.
+ */
+static inline EUSART_HwFlowControl_TypeDef eusart_gecko_cfg2ll_hwctrl(
+	enum uart_config_flow_control fc)
+{
+	if (fc == UART_CFG_FLOW_CTRL_RTS_CTS) {
+		return eusartHwFlowControlCtsAndRts;
+	}
+
+	return eusartHwFlowControlNone;
+}
+
+/**
+ * @brief  Get Zephyr hardware flow control option from
+ *         LL hardware flow control define.
+ * @note   Supports only eusartHwFlowControlCtsAndRts.
+ * @param  fc: LL hardware flow control definition.
+ * @retval UART_CFG_FLOW_CTRL_RTS_CTS, or UART_CFG_FLOW_CTRL_NONE.
+ */
+static inline enum uart_config_flow_control eusart_gecko_ll2cfg_hwctrl(
+	EUSART_HwFlowControl_TypeDef fc)
+{
+	if (fc == eusartHwFlowControlCtsAndRts) {
+		return UART_CFG_FLOW_CTRL_RTS_CTS;
+	}
+
+	return UART_CFG_FLOW_CTRL_NONE;
+}
+
+/**
+ * @brief Main initializer for UART
+ *
+ * @param dev UART device to be initialized
+ * @return int 0
+ */
+static int eusart_gecko_init(const struct device *dev)
+{
+	int err;
+	const struct eusart_gecko_config *config = dev->config;
+	struct eusart_gecko_data *data = dev->data;
+	struct uart_config *uart_cfg = data->uart_cfg;
+
+	EUSART_UartInit_TypeDef eusartInit = EUSART_UART_INIT_DEFAULT_HF;
+	EUSART_AdvancedInit_TypeDef advancedSettings = EUSART_ADVANCED_INIT_DEFAULT;
+
+	/* The peripheral and gpio clock are already enabled from soc and gpio
+	 * driver
+	 */
+	/* Enable EUSART clock */
+	err = clock_control_on(config->clock_dev, (clock_control_subsys_t)&config->clock_cfg);
+	if (err < 0) {
+		return err;
+	}
+
+	/* Init EUSART */
+	eusartInit.baudrate = uart_cfg->baudrate;
+	eusartInit.parity = eusart_gecko_cfg2ll_parity(uart_cfg->parity);
+	eusartInit.stopbits = eusart_gecko_cfg2ll_stopbits(uart_cfg->stop_bits);
+	eusartInit.databits = eusart_gecko_cfg2ll_databits(uart_cfg->data_bits, uart_cfg->parity);
+	advancedSettings.hwFlowControl = eusart_gecko_cfg2ll_hwctrl(uart_cfg->flow_ctrl);
+	eusartInit.advancedSettings = &advancedSettings;
+
+	EUSART_UartInitHf(config->eusart, &eusartInit);
+
+	err = pinctrl_apply_state(config->pcfg, PINCTRL_STATE_DEFAULT);
+	if (err < 0) {
+		return err;
+	}
```

Comment:
Don't you think this should be called before `EUSART_UartInitHf()`?

Reply:
Ok. Like with a lot of these things, I mostly copied the code from `uart_gecko`. This is how it written there.